### ISSUE TITLE
My naiive autoencoder is not learning. Code is very similar to bolts.AE. What Am I missing?

### ISSUE THREAD

**Issue by noamzilo** [PyTorchLightning/pytorch-lightning-bolts#520]
### Before asking:   
1. search the issues.  
I found `pl.bolts.AE`, and I can't see any relevant difference, yet my version doesn't learn anything
 

#### What is your question?

Why isn't my model learning anything? The loss is the same, no matter how many epochs.
This is actually the 1st part of a multi-part question, and I can't even make the simplest version work.

I tried to look at `from pl_bolts.models import AE` but couldn't find what I was missing. Since I don't want just a working AE, but also to understand lightning, I would like to understand my error.

#### Code

```python
import os

import torch
from torch import nn
from torch.nn import functional as F
from torch.utils.data import DataLoader, random_split, Dataset
from torchvision.datasets import MNIST
from torchvision import transforms
import pytorch_lightning as pl
from pytorch_lightning.metrics.functional import accuracy
from pl_bolts.datasets import RandomDataset, DummyDataset, RandomDictDataset
import torch
from pytorch_lightning import LightningModule
from torch.utils.data import Dataset
from pytest import fixture
from Testing.Research.data_modules.CaseDataModule import CaseDataModule
from Testing.Research.config.ConfigProvider import ConfigProvider
from pytorch_lightning.loggers import TensorBoardLogger
from Testing.Research.config.paths import tb_logs_folder
from pl_bolts.datamodules.mnist_datamodule import MNISTDataModule
from pl_bolts.models import AE


class AutoEncoder(LightningModule):
    def __init__(self):
        super().__init__()
        self._input_vector_size = 28 * 28
        self.lr = 0.1

        self.__enc1 = torch.nn.Linear(self._input_vector_size, 128)
        self.__enc2 = torch.nn.Linear(128, 64)
        self.__enc3 = torch.nn.Linear(64, 12)
        # self.__enc4 = torch.nn.Linear(12, 3)
        # self.__dec1 = torch.nn.Linear(3, 12)
        self.__dec2 = torch.nn.Linear(12, 64)
        self.__dec3 = torch.nn.Linear(64, 128)
        self.__dec4 = torch.nn.Linear(128, self._input_vector_size)

    def __encode(self, x):
        x = self.__enc1(x)
        x = F.relu(x)
        x = self.__enc2(x)
        x = F.relu(x)
        x = self.__enc3(x)
        # x = F.relu(x)
        # x = self.__enc4(x)
        return x

    def __decode(self, x):
        # x = self.__dec1(x)
        # x = F.relu(x)
        x = self.__dec2(x)
        x = F.relu(x)
        x = self.__dec3(x)
        x = F.relu(x)
        x = self.__dec4(x)
        return x

    def forward(self, x):
        embedding = self.__encode(x)
        reconstruction = self.__decode(embedding)
        return reconstruction

    def loss(self, orig_batch, reconstruction):
        return F.mse_loss(orig_batch, reconstruction, reduction='mean')

    def __common_step(self, batch, batch_idx):
        orig_batch, label_batch = batch
        reconstruction = self.forward(orig_batch.squeeze().reshape(-1, self._input_vector_size))
        loss = self.loss(orig_batch, reconstruction.reshape(orig_batch.shape))
        return orig_batch, reconstruction, loss

    def training_step(self, batch, batch_idx):
        orig_batch, reconstruction, loss = self.__common_step(batch, batch_idx)
        self.log("train_loss", loss, prog_bar=True, on_step=False, on_epoch=True)
        return loss
        # return {"loss": loss}

    def training_step_end(self, training_step_outputs):
        return training_step_outputs

    # def training_epoch_end(self, outputs) -> None:
    #     torch.stack([x["loss"] for x in outputs]).mean()

    def validation_step(self, batch, batch_idx):
        orig_batch, output, loss = self.__common_step(batch, batch_idx)
        self.log("val_loss", loss, prog_bar=True, on_step=False, on_epoch=True)
        return loss
        # return {"loss": loss}

    # def validation_epoch_end(self, outputs) -> None:
    #     torch.stack([x['loss'] for x in outputs]).mean()

    def test_step(self, batch, batch_idx):
        orig_batch, output, loss = self.__common_step(batch, batch_idx)
        self.log("test_loss", loss, prog_bar=True)
        return loss
        # return {"loss": loss}

    # def test_epoch_end(self, outputs) -> None:
    #     torch.stack([x["loss"] for x in outputs]).mean()

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.parameters(), lr=self.lr)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


@fixture()
def case_datamodule(cfg):
    config = cfg
    real_data_dir_pd = config.real_data_pd_dir
    case_dir = os.path.join(real_data_dir_pd, config.case_dir)
    case_file = os.path.join(case_dir, config.case_file)
    data_module = CaseDataModule(config=config, path=case_file)
    return data_module


@fixture()
def mnist_datamodule():
    return MNISTDataModule()


@fixture()
def tb_logger():
    logger = TensorBoardLogger(save_dir=tb_logs_folder, name='TESTING', default_hp_metric=False)
    return logger


def test_overfit_case_batch(mnist_datamodule, tmpdir, tb_logger):
    datamodule = mnist_datamodule
    # init model
    model = AutoEncoder()

    # Initialize a trainer
    trainer = pl.Trainer(
            max_epochs=2,
            progress_bar_refresh_rate=20,
            overfit_batches=1,
            gpus=1,
            logger=tb_logger,
    )

    # Train the model ⚡
    trainer.fit(model, datamodule=datamodule)

    trainer.test(datamodule=datamodule)

```

   

#### What have you tried?

Looking at the loss in debug, it seems the loss is nonzero.
Thus, I would expect the parameters to change.

#### What's your environment?

 - OS: [e.g. iOS, Linux, Win] Win
 - Packaging [e.g. pip, conda] pip
 - Version [e.g. 0.5.2.1] 1.1.4

**Comment by akihironitta** [comment]
@noamzilo Could I ask what `Testing` package is? In case it's your own module, it would be very helpful if you could provide us with a script without your custom module.

**Comment by noamzilo** [comment]
I'll try to do it when I am more free, can close until then.
